### ISSUE TITLE
Revamp DebugPage in DocService

### ISSUE THREAD

**Issue by ikhoon** [line/armeria#2844]
DocService is a loved feature by Armeria users.
While we are preparing to [release 1.0](https://github.com/line/armeria/milestone/151), we don't have enough time to improve the debug console.
One of our contributors tried to revamp the console, however, it is stale.😭 #2194 #1954 #2118

A full-featured autocomplete editor is not easy to implement by ourselves. We are not a front-end specialist.

I think https://github.com/fullstorydev/grpcui#request-form will be able to provide convenient functions to users without being too complicated. Users might easily fill the table for the JSON content without prior knowledge of a data structure. #2690

**Comment by ikhoon** [comment]
Related work: Another open-source written in Go that provides better design for API documentation: https://github.com/encoredev/encore#api-documentation

**Comment by Dogacel** [comment]
I think `grpcui` has a beautiful UI but they use go templates, so I doubt we can utilize their code right away. We have to rebuild lots of things. I think it is pretty inspiring.

I suggest using the well-known [monaco-editor](https://microsoft.github.io/monaco-editor/) inside our debug form to:
- enable auto-complete
- show syntax errors
- syntax highlighting + more themes (🥳)

Monaco editor supports JSON Schema validation out of the box. [**Demo found from web here**](https://codesandbox.io/s/gue0q)

My plan is to modify `ServiceSpecification` a little bit to achieve generate a JSON schema. I have seen some tools to generate JSON schema ([Example](https://github.com/chrusty/protoc-gen-jsonschema)) but I am not 100% sure how we can utilize that in Armeria.

@ikhoon any suggestions?

**Comment by minwoox** [comment]
> I suggest using the well-known [monaco-editor](https://microsoft.github.io/monaco-editor/) inside our debug form to:

That sounds good. 👍 

> I have seen some tools to generate JSON schema ([Example](https://github.com/chrusty/protoc-gen-jsonschema)) but I am not 100% sure how we can utilize that in Armeria.

It seems like there's no JVM implementation of Gradle plugin for the schema compiler. So we should probably create Gradle plugin that creates `JSONSchemas` using go, then we can apply it.
I haven't received any demand for the validation of it, so how about just applying the editor first and then moving on?

**Comment by Dogacel** [comment]
> It seems like there's no JVM implementation of Gradle plugin for the schema compiler. So we should probably create Gradle plugin that creates `JSONSchemas` using go, then we can apply it.

I have no experience with this, do you have any sources or examples? I found https://github.com/gogradle/gogradle but seems like it is pretty old.

**Comment by minwoox** [comment]
> I found [gogradle/gogradle](https://github.com/gogradle/gogradle) but seems like it is pretty old.

Yeah, it's pretty old and not maintained. We have used it in our other project before but we completely removed that due to the reason.
https://github.com/line/centraldogma/pull/707/files#diff-de8fe44d5ac06e1637fb405beaa31259f0a95163399c0f8cb4158c820b9efcd0L20

We have these options:
- Creating protoc-gen-jsonschema gradle plugin
- Creating the gogradle plugin equivalent and using protoc-gen-jsonschema compiler
- Using the existing Gradle plugin for Golang that might lack some features (e.g. https://plugins.gradle.org/plugin/org.curioswitch.gradle-golang-plugin)
  - We could add the feature we need and make it work

All these options are not easy so we probably need to spend a lot of time.
But if you really wish to add it, then let's start to discuss the options. 😄 
By the way, we have other issues such as #1670, #90 just in case you are interested in. 😉

**Comment by ikhoon** [comment]
> My plan is to modify ServiceSpecification a little bit to achieve generate a JSON schema. I have seen some tools to generate JSON schema ([Example](https://github.com/chrusty/protoc-gen-jsonschema)) but I am not 100% sure how we can utilize that in Armeria.

It is an excellent idea but would it be better to directly convert `ServiceSpecification` into `JSON schema` rather than using `protoc-gen-jsonschema` so that other protocols such as Thrift or annotated services take advantage of it?

**Comment by Dogacel** [comment]
> > My plan is to modify ServiceSpecification a little bit to achieve generate a JSON schema. I have seen some tools to generate JSON schema ([Example](https://github.com/chrusty/protoc-gen-jsonschema)) but I am not 100% sure how we can utilize that in Armeria.
> 
> It is an excellent idea but would it be better to directly convert `ServiceSpecification` into `JSON schema` rather than using `protoc-gen-jsonschema` so that other protocols such as Thrift or annotated services take advantage of it?

Makes sense, let me start experimenting with it 😄
